Feat/first nib images

mateipopa opened the pull request:
This is still work in progress, but while there still are things to test, it'll give you an idea of what the approach is.
Please fire away with feedback - it'll be appreciated as usual.

mjr-blockjoy commented on ethereum/ethereum-erigon/babel.yaml:
```diff
@@ -0,0 +1,111 @@
+  - key: erigon-goerli-archive
+    archive_pointers:
+      - pointer: !store_id ethereum-erigon-goerli-archive-v1
+      - pointer: !store_id ethereum-erigon-goerli-archive-v1 
```
It doesn't make sense to define `new_archive: true` property if it have the same store_id.

mateipopa replied:
What I wanted to achieve here is that whenever you select the tracing API, the dataset should point to the archive store rather than the full one. Can we configure this behavior more explicitly?

calebcall commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,172 @@
+            GRAFANA_PROM_API_KEY=${{ secrets.GRAFANA_PROM_API_KEY }}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          tags: ghcr.io/blockjoy/${{ steps.version.outputs.image_name }}:${{ steps.version.outputs.image_tag }}
```
I think we should also include a latest tag as we previously discussed.  Trying to update all the dependencies with the hash is going to create a lot of work for every image update.  Or we omit the hash and stick to the semantic version (i.e. we shouldn't ever have a version 1.1.1-hask8d and a 1.1.1-ja9dk1...two of the same sematic with different hashes)

mateipopa replied:
Agreed, this mainly stemmed from the fact that I needed a way to bootstrap the pipeline since initially, no images were built that could satisfy the requirements of the next images. I ended up going for the commit of the hash and then updated the dependent Dockerfiles manually. We should only need the semver and only go for full release versioning, so don't allow pre-release versions on prod.
I'm not yet sure where we should draw the line between automations and manual deployments, especially to prod. Namely, we'll have renovate update Dockerfiles when new images get build (this is still something I need to test, haven't done it yet) and then CI/CD will pick up on those changes and built the three groups of images (node-base <-- clients <-- protocols) and (either renovate or CI/CD, TBD) will update the docker-container:// URI in the babel.yaml, but do we want afterwards to deploy the image to the API? I tend to think we do, so that it will become available on the frontend and then we could select what nodes should run what images independently

calebcall commented on clients/erigon/Caddyfile.template:
```diff
@@ -0,0 +1,41 @@
+}
+
+http://{{ hostname }}{{ tld }}, https://{{ hostname }}{{ tld }}, *.blkjy.io, http:// {
+	log {
```
```suggestion
	tls {
		dns cloudflare {$CLOUDFLARE_API_KEY}
		resolvers 1.1.1.1
	}
	log {
```

calebcall replied:
We should include this, this will prevent the nodes from trying to use 127.0.0.1 for dns lookups, which doesn't work and helps gets certs.

mateipopa replied:
Also there's a typo in the variable syntax

calebcall commented on clients/reth/Caddyfile.template:
```diff
@@ -0,0 +1,41 @@
+}
+
+http://{{ hostname }}{{ tld }}, https://{{ hostname }}{{ tld }}, *.blkjy.io, http:// {
+	log {
```
```suggestion
	tls {
		dns cloudflare {$CLOUDFLARE_API_KEY}
		resolvers 1.1.1.1
	}
	log {
```

calebcall commented on ethereum/ethereum-reth/main.rhai:
```diff
@@ -0,0 +1,112 @@
+    services: [
+        #{
+            name: "reth",
+            run_sh: `export HOME=/root && /root/bin/reth node \
+                    --datadir=${global::HOME} \
+                    --chain=${global::NET} \
+                    --http \
+                    --http.addr=127.0.0.1 \
+                    --http.port=${global::RPC_PORT} \
+                    --http.api=eth,net,web3,debug,trace,txpool \
+                    --ws \
+                    --ws.addr=127.0.0.1 \
+                    --ws.port=${global::WS_PORT} \
+                    --ws.api=eth,net,web3,debug,trace,txpool \
+                    --metrics=127.0.0.1:${global::METRICS_PORT} \
+                    --authrpc.jwtsecret=${global::HOME}/jwt.hex \
+                    ${global::RETH_EXTRA_ARGS}`,
```
I think we should discuss doing the newline for each parameter.  It makes trying to troubleshoot the running processes really challenging.  Maybe we could do something in bv to remove them when creating the jobs?

mateipopa replied:
Agreed on this, it makes the RHAI more readable, but in the end the job config becomes a mess. I don't think it would be too much work, but this one's for @mjr-blockjoy

calebcall commented on ethereum/ethereum-reth/babel.yaml:
```diff
@@ -0,0 +1,93 @@
+      ports:
+        - port: 30303
+          name: P2P port
+        - port: 42069
```
Reth doesn't have torrent support

mateipopa replied:
Yeah, I kind of rushed through the reth image (or rather Windsurfed) in order to get it to build so I could put the CI/CD pipeline together. I'll go through the finetuning soon

calebcall commented on ethereum/ethereum-reth/main.rhai:
```diff
@@ -0,0 +1,112 @@
+                    --http \
+                    --http.addr=127.0.0.1 \
+                    --http.port=${global::RPC_PORT} \
+                    --http.api=eth,net,web3,debug,trace,txpool \
```
We defined tracing in the babel.yaml file as a configurable value but then it's hardcoded here.

calebcall reviewed (commented):
Added several comments.  Most of this, from what I know of the new process, looks great.

mateipopa commented:
> Added several comments. Most of this, from what I know of the new process, looks great.

Have a look at the github actions too please - I've mostly focused on the CI/CD part until now, will get to the actual image fine-tuning a bit later. 

The CI/CD workflow is building a matrix based on the location of files changed. If the files get on the PR compared to the source branch, the image gets pulled in and added to groups. There are three groups so far, the base, clients and protocols. Each group has internal paralellization, but can't run before the groups of images they depend on so we don't break dependencies.
The approach so far is to build the images when a PR gets created (so no pushing on main without PR), enforce checks and ultimately require reviews before the PR can be merged.
From my point of view, this should be fine because when you're developing the image locally, you should have all the tooling to ensure the docker images build properly before you actually commit and open a PR, so the confidence it will work should be pretty high. Once the PR is opened, the images start to build in order to test, so this is the question: 
In order for the final protocol level image to be able to build, it needs the lower level images to already be build so they can reference them in the FROM. Do we want to build the images at this point or do we only want to build the images when the PR is merged so we save on building resources and just have some basic checks on the Dockerfile (probably linting and the likes). 
AFAICT, building early shouldn't be an issue, especially after we'll configure our own GHA runners that will run the builds since the artifacts (or at least layers) will already be there when merging to main. We should probably add some cleanup routine for images that get built from PRs that ultimately don't get merged. Or have pre-release versioning for PR based images and cleanup those, while the merged PRs will result in release images that we can then use in production? We could also remove the images immediately after we validate the build works properly. 
We need to also take into account that this process would get triggered automatically by renovate once it opens PRs to update dependencies.

On another note, I've added some logic on the GHA workflow to retrieve the client version in the Dockerfile (it looks like ERIGON_VERSION or RETH_VERSION) using some regex and adds that to the client docker image tag so we know what's actually running inside (not sure if we'll want to stick with this in the end: what version would we retrieve when more than one component is running inside the image? TBD). There is the question about the opportunity to have renovate update the client versions inside the dockerfiles too, but I've refrained from doing it because it seems like it could get out of control. We might be able to pull it off with minor versions AND some more extensive tests (we'd somehow need to determine if the new client actually works with arguments configured in the RHAI) to be sure the automated update doesnt break the image. 

Also, I'm not yet sure where we should draw the line between automations and manual deployments, especially on prod. Namely, we'll have renovate update Dockerfiles when new images get build (this is still something I need to test, haven't done it yet) and then CI/CD will pick up on those changes and built the three groups of images (node-base <-- clients <-- protocols) and (either renovate or CI/CD, TBD) will update the docker-container:// URI in the babel.yaml, but do we want afterwards to deploy the image to the API? I tend to think we do, so that it will become available on the frontend and then we could select what nodes should run what images in a controlled fashion.

mjr-blockjoy commented on ethereum/ethereum-reth/babel.yaml:
```diff
@@ -0,0 +1,93 @@
+        impact:
+          new_archive: true
+          add_cpu: 3
+          add_memory_bytes: 10000000
```
This is already changed, in latest `nib` version:
```
add_memory_mb: 100
add_disk_gb: 30
```

mateipopa commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,172 @@
+
+jobs:
+  detect-changes:
+    runs-on: ubuntu-latest
```
Agreed, I think the best approach here would be to use semver compatible versions and let renovate or the CI/CD workflow handle the dependencies

mateipopa commented on node-base/templates/netdata-stream-conf.template:
```diff
@@ -0,0 +1,10 @@
+[stream]
+    enabled = yes
+    destination = tcp:netdata.blkjy.io
+    api key = 894aa814-e672-4df1-9e8c-3bc919f1b905
```
No, I'll add this as a GHA secret too

calebcall replied:
We can actually remove the netdata config stuff for inside the containers now.

mateipopa commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,172 @@
+    needs: [detect-changes, build-base]
+    if: |
+      always() &&
+      needs.build-base.result != 'failure' &&
```
Possibly, will check

mateipopa commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,172 @@
+        run: |
+          SHA=$(git rev-parse --short HEAD)
+          CLIENT_NAME=$(basename ${{ matrix.image_path }})
+          IMAGE_NAME="blockjoy-${CLIENT_NAME}"
```
The namespace is already /blockjoy here because that's the organization's name, I added blockjoy- again to distinguish it from a potential vanilla erigon image

calebcall replied:
I'm not a fan of `blockjoy/blockjoy-${CLEINT_NAME}`.  I like the intent though. Are customers going to be pushing their images to the blockjoy namespace?  If not then the extra is redundant.  Maybe we use a different namespace for community created images.  Otherwise there's nothing stopping someone (other than us) from creating `blockjoy/blockjoy-malicious-image`.  Or overwriting `blockjoy/blockjoy-reth` with a malicious or bad image.

calebcall commented on clients/erigon/config-alloy.template:
```diff
@@ -0,0 +1,206 @@
+
+	endpoint {
+		name = "prom-ep"
+		url  = "https://prometheus-prod-24-prod-eu-west-2.grafana.net/api/prom/push"
+
+		basic_auth {
+			username = "1322871"
+			password = "${GRAFANA_PROM_BASICAUTH}"
+		}
```
This needs to be updated to the new metrics endpoint.  You can see the node-setup script for the new values.

calebcall commented on clients/reth/config-alloy.template:
```diff
@@ -0,0 +1,111 @@
+
+prometheus.remote_write "default" {
+        endpoint {
+                url = "https://prometheus-prod-24-prod-eu-west-2.grafana.net/api/prom/push"
+
+                basic_auth {
+                        username = "1234567"
+                        password = "${GRAFANA_PROM_BASICAUTH}"
+                }
```
Same as the erigon image.  Needs the new metrics endpoint/username/password

calebcall reviewed (commented):
Few things to talk though, not extremely important though.
